Adds Kernel.defguard/1 macro to build guard-safe macros

christhekeele opened the pull request:
Based on previous feedback (#5854), this is an alternate `defguard` implementation that moves all of the heavy lifting to `:elixir_expand.expand/2`.

Most of the code is raising helpful `ArgumentErrors` to direct the user towards proper usage, and actually generating the guard code.

In between the two, we make a copy of the `__CALLER__` env, give it a guard context and access to any variables specified the guard's parameter list, and use it to expand the guard implementation:

`:elixir_expand.expand(impl, %{ __CALLER__ | context: :guard, vars: vars}`

If there is anything fishy with `impl`, including guard-specific syntax and semantic conditions, this will raise a `CompileError`.

Closes #2469.

josevalim commented on lib/elixir/lib/macro.ex:
```diff
@@ -298,6 +298,86 @@ defmodule Macro do
+        end
+      end
+  """
+  @spec guard(Macro.t, [Macro.t]) :: Macro.t | no_return
```
`Kernel.Utils` may be a better place for putting this, since I don't believe this should be public.

josevalim commented on lib/elixir/test/elixir/macro_test.exs:
```diff
@@ -620,6 +621,57 @@ defmodule MacroTest do
 
+  # guard
+
+  test "guard" do
```
It would be nice to test the error messages too and break this test into multiple ones (describe can help!).

christhekeele replied:
I meant to ask about that--I'm struggling to catch those syntatic `ArgumentErrors` and semantic `CompileErrors` in ExUnit; neither `catch_error` or `assert_error` are up to the task since the errors are compile-time.

christhekeele replied:
i.e.:

```elixir
# lib/elixir/test/elixir/kernel/guard_test.exs
Code.require_file "../test_helper.exs", __DIR__

defmodule Kernel.GuardTest do
  use ExUnit.Case, async: true

  # Kernel.defguard usage validation

  test "defguard requires a proper name" do
    assert_raise ArgumentError, fn ->
      defguard "asdf"
    end
  end

  # Kernel.defguard implementation validation

  test "defguard fails to compile expressions not allowed in guard" do
    assert_raise CompileError, fn ->
      defguard foo(bar) when send self(), "message"
    end
  end

end
```

---

`$ bin/elixir lib/elixir/test/elixir/kernel/guard_test.exs`

> ** (ArgumentError) could not extract a definition for defguard from `"asdf"`, expected: `defguard name(args...) when implementation...`
>     (elixir) lib/elixir/lib/kernel.ex:4030: Kernel."MACRO-defguard"/3
>     (elixir) expanding macro: Kernel.defguard/1
>     lib/elixir/test/elixir/kernel/guard_test.exs:10: Kernel.GuardTest."test defguard requires a proper name"/1

> ** (CompileError) lib/elixir/test/elixir/kernel/guard_test.exs:18: cannot invoke remote function :erlang.send/2 inside guard
>     (elixir) lib/elixir/lib/kernel.ex:4072: Kernel.do_defguard/5
>     (elixir) expanding macro: Kernel.defguard/1
>     lib/elixir/test/elixir/kernel/guard_test.exs:18: Kernel.GuardTest."test defguard fails to compile expressions not allowed in guard"/1

christhekeele replied:
(Incidentally I'm not sure if it's idiomatic to wrap code in error messages in backticks in Elixir, but this shows why I normally do: it helps it jump out in pasted stacktraces. Will remove on request.)

josevalim commented on lib/elixir/lib/kernel.ex:
```diff
@@ -3994,6 +3993,97 @@ defmodule Kernel do
+    end)
+    refs = for {ref, _meta, context} <- vars, do: {ref, context}
+    test_env = %{env | context: :guard, vars: env.vars ++ refs}
+    {_valid?, _env} = :elixir_expand.expand(implementation, test_env)
```
Beautiful!

christhekeele replied:
Nothing is prettier than -200 lines of code :)

josevalim commented on lib/elixir/lib/kernel.ex:
```diff
@@ -3994,6 +3993,97 @@ defmodule Kernel do
+        {node, acc}
+    end)
+    refs = for {ref, _meta, context} <- vars, do: {ref, context}
+    test_env = %{env | context: :guard, vars: env.vars ++ refs}
```
I think `env.vars` here should only be refs. After all, `defguard` starts a new variable scope.

josevalim commented on lib/elixir/test/elixir/kernel/guard_test.exs:
```diff
@@ -0,0 +1,111 @@
+  # end
+  #
+  # describe "Kernel.defguard compilation" do
+  #   test "fails on expressions not allowed in guard" do
```
You need to use defmodule inside the test and define the guard inside that module.

christhekeele commented:
@kelvinst worth noting, the current failing test is intentionally checked in, as there is a situation this PR doesn't yet fully solve to my satisfaction.

christhekeele commented:
Bumping:

After looking at this again, I'm still not sure about the best way to handle my remaining misgiving.

`defguard foo(local, call) when local.(call)` compiles when it shouldn't, because it doesn't raise red flags in `:elixir_expand.expand(guard, env)`.

The charm of this implementation, of course, is that it entirely relies on `:elixir_expand.expand`. 😩 What to do?

Back in May I got very deep (out of my depth) into the compiler trying to find the right incantation of the right `:elixir_` function to catch this case, and taking another glance at it I don't think I'm liable to make progress anytime soon.

In my dream world the snippet above would raise with a `CompileError: invalid expression in guard` since that's what happens when you try to compile the literal result of the macro.

It's worth asking, though: are we okay with this incongruity, where instead you get a runtime `CompileError: illegal guard expression` on usage, for the sake of keeping the implementation lean?

Otherwise, I'm not sure how to progress, and am open for pointers. 😃

josevalim commented:
@christhekeele the compiler should raise if `x.(args)` is used in guards.

josevalim commented:
@christhekeele I found other places we were not properly checking for this, so I will push a PR soon. Thanks for the heads up!

christhekeele commented:
@josevalim Interesting--in my testing I had to do the [same thing](https://github.com/elixir-lang/elixir/pull/5857/files#diff-3c8e0f951ee365844de7eb64812cf62dR108) with comprehensions to get it to raise––maybe a related (or identical) issue?

josevalim commented:
@christhekeele yup, all fixed in master and v1.5.

christhekeele commented:
Cool--rebased, squashed, crafted a better commit message, and updated the suite, works great! Now that it isn't half-implemented anymore, you can revisit the hard questions at leisure:

> We still have not decided if we are going to merge this feature or not.

😸 Either way I'm very satisfied I was able to get the last fiddly bits of this PR to work.

christhekeele commented:
After typing tons of `defguard` formulations in the last––looks like 10 months since I first started playing with this––paired with these error updates, I recant my previous stance in the discussion on #2469.

The `defguard name(args..) when impl...` syntax just ***feels*** right after typing it a bunch.

`defguard` has two chief limitations:

- it must be a one-liner
- it must only use other guardables

My chief argument for `do... end` was that people are going to impulsively type that a lot no matter how well they know the API. There are several block-less `defxxx` macros, but they mostly take data rather than code as their arguments. My worry was that even if you know `defguard` really well, you'll still instinctively try to wrap the code in a `do` block since that's how pretty much everything else in Elixir works.

I was right, too. Even after implementing the damn thing and playing with it for months, I still accidentally throw it in a `do` block all the time. 😆 

With the right errors in place, though, you figure it out pretty quickly. The `undefined function defguard/2` you get might be a little confusing if you've never written macros before, but if you've played with them at all you grok very quickly that it doesn't accept a block since that feedback is consistent with what you've seen before.

That prompts you to remember it takes a `when` formulation. The interesting thing I noticed is the following: I **never**, **ever** accidentally tried to use non-guardables or multiple expressions when starting to write a guard with the correct `when` formulation, because you associate the limitations with *using* guards instinctively when *writing* them in that format. And that comes a lot earlier in learning Elixir than groking the implications of `undefined macro defguard/2`.

However, when I sat down to write a new `defguard` but started off wrong with a block, about half the time I went to fix that first `defguard/2` compile error I noticed I had let little unguardables and the occasional multi-line phrase sneak into the implementation. The process of reformulating it into a `when` expression immediately made me notice **those** errors well before the compiler had to alert me to them.

Essentially, the high-fidelity `invalid expression in guard` error messages are great, but almost unneeded because starting off in the right context––`when`––is over half the battle. Good call. 😃 👍

josevalim commented:
Thanks for the feedback @christhekeele!

We still haven't decided on how to proceed with this. We are exploring other venues that may potentially make this obsolete but it is still way too early to tell. I will have this assigned to Elixir 2.0 milestone since, by then, the other approaches will either have succeeded or failed (which means we should likely include this in).

alco commented on lib/elixir/lib/kernel.ex:
```diff
@@ -4117,6 +4117,64 @@ defmodule Kernel do
+
+        def is_even_and_small?(value) do
+          if is_even(value) and value <= 100, do: true, else: false
+        end
```
Not the best examples, they can simplified to

```elixir
def is_even_and_large?(value), do: is_even(value) and value > 100
def is_even_and_small?(value), do: is_even(value) and value <= 100
```

A better example (https://en.wikipedia.org/wiki/Collatz_conjecture):

```elixir
@doc """
The Collatz conjecture is a conjecture in mathematics named after Lothar Collatz. It can be
summarized as follows. Take any positive integer n. If n is even, divide it by 2 to get n / 2. If n
is odd, multiply it by 3 and add 1 to obtain 3n + 1. Repeat the process indefinitely. The conjecture
is that no matter what number you start with, you will always eventually reach 1.

This function takes a positive integer number and returns the number of steps it takes to reach 1.
"""
def collatz(n) when n > 0, do: collatz(n, 0)

defp collatz(1, step_count), do: step_count
defp collatz(n, step_count) when is_even(n), do: collatz(div(n, 2), step_count + 1)
defp collatz(n, step_count), do: collatz(3*n + 1, step_count + 1)
```

christhekeele replied:
:) What a great example, I love the Collatz conjecture!

christhekeele replied:
The examples I provided are awkward, but try to concisely showcase that `is_even/1` can be used both inside and outside guards. I don't think there's an opportunity to do both when implementing collatz, though. Definitely needs a better example though.

alco replied:
Having 2 separate examples would work just as well. You could add a paragraph saying

> Just like the builtin the guards, the ones defined with `defguard` can also be used like normal functions:
>
> ```elixir
> if is_even(num), do: <something>
> ```

christhekeele commented:
Out of curiosity, do you have any links I could follow to learn more about

> other venues that may potentially make this obsolete

I don't see anything in the mailing lists but am intrigued.

josevalim commented on lib/elixir/test/elixir/kernel/guard_test.exs:
```diff
@@ -0,0 +1,361 @@
+Code.require_file "../test_helper.exs", __DIR__
+
+defmodule Kernel.GuardTest do
+  use ExUnit.Case, async: true
```
You can't use `capture_io(:stderr)` with `async: true`. My recommendation would be to move those warnings to "test/kernel/warnings_test.exs".

christhekeele replied:
Are the examples there the preferred way to write tests that test compiling code? ie:

```elixir
  test "unused variable in defguard" do
    capture_err(fn ->
      Code.eval_string """
      defmodule Sample do
        defguard foo(bar, baz) when bar
      end
      """
    end) =~ "variable \"baz\" is unused"
  after
    purge Sample
  end
```

josevalim replied:
Those are testing warnings. For testing errors, you can probably use assert_raise:

```elixir
assert_raise error, message, fn ->
  defmodule Foo do
    defguard ...
  end
end
```

josevalim commented on lib/elixir/lib/kernel.ex:
```diff
@@ -4295,6 +4295,65 @@ defmodule Kernel do
+  end
+
+  defp do_defguard(_call, name, args, _impl = [], env) do
+    message = :elixir_def.format_error({:bodyless_clause, :defguard, {name, length(args)}})
```
I would say we should just expand this case to:

```elixir
defmacro unquote(call)
```

And Elixir will take care of emitting the warning if no implementation is defined later.

josevalim commented:
Hi @christhekeele! We finally plan to include this feature in Elixir v1.6 and I have added two comments, one of them should fix the build.

One last thing to consider is this case:

    defguard is_foo(bar) when is_this(...) when is_that(...)

I think we should raise in those cases. Only one `when` is allowed since we cannot reliably support two `when`s if the guard is used inside an `and` or an `or`.

OvermindDL1 commented:
Still no chance of adding in structure matching support like my [defguard](https://github.com/overminddl1/defguard) hack of an example library does (it should be built into the `def` call for in-place expansion like I do)?

josevalim commented:
Hi @OvermindDL1, [as asked on #2469](https://github.com/elixir-lang/elixir/issues/2469#issuecomment-285395933), if you would like to see a different behaviour, please send a proposal to the mailing list.

OvermindDL1 commented:
Done at:  https://groups.google.com/forum/#!topic/elixir-lang-core/tu-0yjvB0Kk
Also see past discussion (or lack there-of really, only comments about it was someone saying how they would love such functionality) at:  https://elixirforum.com/t/defguard/4052

josevalim commented:
Btw, we also need `defguardp/1`. :)

josevalim commented:
@christhekeele I am also wondering if we should force all arguments to be variables or a `var \\ default`. The reason is that people will try to match inside `defguard` and right now that would translate to a compile time matching, which is most likely not what they want.

OvermindDL1 commented:
> The reason is that people will try to match inside defguard and right now that would translate to a compile time matching, which is most likely not what they want.

I'm definitely for this.  If the functionality is extended later this would make it backwards-compatible.

christhekeele commented:
- [ ] fix specs
- [ ] `defguardp`
- [ ] allow compiler to do the complaining about bodiless clauses

Discussion:

- [ ] handle double `when`

Concerning double `when`, I didn't even realize that was valid syntax, though it makes sense from an erlang perspective. The current behaviour for macros and functions is effectively to `or` them together, in my experimentation, is that right? `:elixir_utils.extract_guards(guard)` does return both clauses, but the `defguard` implementation doesn't make proper use of them, could we mimic the usual behaviour by manually`or`ing them together? I suspect you already answered this question but I don't quite get what you meant by

> Only one when is allowed since we cannot reliably support two `when`s if the guard is used inside an `and` or an `or`.

- [ ] handle matching in `defguard`

Still playing with this one, though I'll probably have questions later. :)

josevalim commented:
> could we mimic the usual behaviour by manuallyoring them together?

`when` is not the same as `or`. An error on the left side of a `when` does not affect the right side, but an error on the left side of an `or` cause the whole operator to fail. Because of that, we can't really translate a `when` to an `or` because on code like this: `when is_atom(foo) and custom_guard(...)` the only way we could expand `custom_guard` is by converting it to multiple `when`s and I would like to avoid that (at least for now).

josevalim commented:
And regards to "handle matching in defguard", we are not adding  this feature on v1.6, so we can postpone it. For now it is enough to guarantee that the arguments are `variables` or `var \\ default`. `defdelegate` does something similar.

christhekeele commented:
Ah, makes sense. Following the thread [here](https://groups.google.com/d/msg/elixir-lang-core/tu-0yjvB0Kk/uamy8BY6AgAJ) also helped me see the nuances.

By "handle matching" I just mean "warn or error on matching, based on feedback". The thread does help me realize this should be an error.

christhekeele commented:
Added some specs to signify an issue I noticed: in the current implementation you cannot use other funs/macros previously defined in a module in a `defguard` macro. I suspect the `:elixir_expand.expand/2` call doesn't have a quite correctly set up `Macro.Env` but I won't be able to look at it for a few days at least.

- [ ] Fix that thing

josevalim commented on lib/elixir/lib/kernel.ex:
```diff
@@ -4295,6 +4295,95 @@ defmodule Kernel do
+            raise ArgumentError, message: "invalid syntax in defguard #{Macro.to_string call}"
+        end
+      {call, _multiple_impls} ->
+        raise ArgumentError, message: "invalid syntax in defguard #{Macro.to_string call}"
```
Here we could say in the error message that only a single `when` is allowed

josevalim replied:
Btw, you don't need the `message: ` thingy. `raise ArgumentError, "foo bar"` is enough.

josevalim commented on lib/elixir/lib/kernel.ex:
```diff
@@ -4295,6 +4295,95 @@ defmodule Kernel do
+        case Macro.decompose_call(call) do
+          {_name, args} ->
+            validate_variable_only_args!(call, args)
+            {^args, refs} = extract_refs_from_args(args)
```
This and the line above could be done in one pass now. One that validates and extracts.

josevalim commented:
@christhekeele thanks for all of the updates :heart:. I took a quick look at the code and indeed it is not clear why `defguard` does not work with previous macros `defguard` defined in the same code. If you need help in figuring that out please let us know - but there is no worry. The next Elixir release is coming out only in January/2018. :)

christhekeele commented:
Ah good to know, should give me ample opportunities to suss it out myself then, and will have plenty time to call in reinforcements if need be and tidy it up. Cheers!

josevalim commented:
This has been merged manually, thank you @christhekeele! Great work! :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

I have done two changes:

  1. I have added defguard to the formatter and formatted your code

  2. The failing test was happening because by the time we expanded `defguard`, none of the other macros have indeed been defined. I fixed this by making the body of the defguard macro be a call to another macro. By the time the macro in the body is expanded, the other macros in the same module will be available

christhekeele commented:
😍 Super cool news! Thanks for all the support in reaching this implementation.
